### ISSUE TITLE
Assertion failed at mrb_field_write_barrier()

### ISSUE THREAD

**Issue by matsumoto-r** [mruby/mruby#2996]
The following code occurred abort() with `conf.enable_debug`  on both Mac OSX and Linux(ubunt14.04 x86_64).

- build_config.rb

```ruby
MRuby::Build.new do |conf|
  toolchain :gcc
  conf.gembox 'default'

  conf.gem :mgem => 'mruby-json'
  conf.gem :mgem => 'mruby-io'

  conf.enable_debug
end
```

but, if `conf.enable_debug` was removed, abort() don't occur.

```
$ ./bin/mruby reproduce.rb
Assertion failed: (!(((value)->color & ((mrb)->current_white_part ^ (1 | (1 << 1))) & (1 | (1 << 1))) || (value)->tt == MRB_TT_FREE) && !(((obj)->color & ((mrb)->current_white_part ^ (1 | (1 << 1))) & (1 | (1 << 1))) || (obj)->tt == MRB_TT_FREE)), function mrb_field_write_barrier, file /Users/matsumoto_r/DEV/mruby/src/gc.c, line 1156.
Abort trap: 6
```

```ruby
$ cat reproduce.rb 
def data_each total_line
  File.open("resource.log", "r") do |file|
    read_line = 0
    file.each do |line|
      yield JSON.parse(line)
      read_line += 1
      break if read_line > total_line
    end
  end
end

def analyze_data d, key, res
  res[d[key]] = 0 if res[d[key]].nil?
  res[d[key]] += 1
end


result = {}
data_each 10000 do |line|
  analyze_data line, "hostname", result # => Assertion failed at 5000 loop
end
```

`resource.log` is https://github.com/matsumoto-r/rcheck-analyzer/blob/master/test/resource.log

**Comment by Asmod4n** [comment]
Looks like there are some changes to mruby master, tried that version too?

**Comment by matsumoto-r** [comment]
mruby version is f07ee20260156e07c2ca5efe63bf0ea9f3ac8132 .

**Comment by cremno** [comment]
Might be related to https://github.com/mruby/mruby/issues/2825.

**Comment by matz** [comment]
I fixed #2825 by fc4f0b9.  could you check if this issue is fixed?

**Comment by matsumoto-r** [comment]
I just now check this commit, and this issue is fixed! Thank you for quick response!
